File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let [l,r,ans] = [0,0,0];
    const visited = new Set();
    while(r<s.length){
        if(visited.has(s[r])){
            ans = Math.max(ans,r-l);
            // update window
            while(visited.has(s[r])){
                visited.delete(s[l]);
                l++;
            }
        }
        visited.add(s[r]);
        r++;
    }
    return Math.max(ans,r-l);
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l = 0, r = 0;
        int ans = 0;
        unordered_set<char> visited;
        while(r<s.size()){
            if(visited.find(s[r]) != visited.end()){
                ans = max(ans,r-l);
                // update window 
                while(visited.find(s[r]) != visited.end()){
                    visited.erase(s[l]);
                    l++;
                }
            }
            visited.insert(s[r]);
            r++;
        }
        return max(ans,r-l);
    }
};